homeassistant/components/konnected/config_flow.py

```diff
@@ -0,0 +1,732 @@
+            return self.hass.data[DOMAIN]["config_data"].get(self.device_id)
+        return None
+
+    @callback
+    def async_cache_config(self, config):
+        """Cache the device config for shared usage."""
+        if DOMAIN not in self.hass.data:
+            self.hass.data[DOMAIN] = {}
+            self.hass.data[DOMAIN]["config_data"] = {}
+
+        self.hass.data[DOMAIN]["config_data"][self.device_id] = config
+
+    async def async_step_init(self, user_input=None):
+        """Needed in order to not require re-translation of strings."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_import(self, import_info):
+        """Import a configuration.yaml config.
+
+        This flow is triggered by `async_setup` for configured panels.
+        Triggered for any panel that does not have a config entry
+        yet (based on device_id).  If the cfg entry can't be made
+        we'll cache the data in hass.data[DOMAIN]["config_data"] for
+        other flows to reference.
+        """
+        _LOGGER.debug(import_info)
+        try:
+            device_config = DEVICE_SCHEMA_YAML(import_info)
+
+        except vol.Invalid as err:
+            _LOGGER.error(
+                "Cannot import config..%s", humanize_error(import_info, err),
+            )
+            return self.async_abort(reason="unknown")
+
+        # swap out pin for zones in a io config
+        def pins_to_zones(config):
+            for zone in config:
+                if zone.get(CONF_PIN):
+                    zone[CONF_ZONE] = PIN_TO_ZONE[zone[CONF_PIN]]
+                    del zone[CONF_PIN]
+
+        if device_config.get(CONF_BINARY_SENSORS):
+            pins_to_zones(device_config[CONF_BINARY_SENSORS])
+
+        if device_config.get(CONF_SENSORS):
+            pins_to_zones(device_config[CONF_SENSORS])
+
+        if device_config.get(CONF_SWITCHES):
+            pins_to_zones(device_config[CONF_SWITCHES])
+
+        self.device_id = self.context["device_id"] = device_config["id"]
+        try:
+            self.host = device_config[CONF_HOST]
+            self.port = device_config[CONF_PORT]
+
+        except KeyError:
+            # cache config and wait for user input or discovery to provide host info
+            self.async_cache_config(device_config)
+            return await self.async_step_user()
+
+        # create the config entry
+        return await self.async_create_or_update_entry(device_config)
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input:
+            try:
+                self.host = user_input[CONF_HOST]
+                self.port = user_input[CONF_PORT]
+
+                # try to obtain the mac address from the device
+                status = await get_status(self.hass, self.host, self.port)
+                self.device_id = status.get("mac").replace(":", "")
```

Comment:
This will error if the dict lookup returns `None`. If we know that they key is in the dict we should use `dict[key]`.

Reply:
👍